orchagent/debugcounterorch.cpp

```diff
@@ -0,0 +1,589 @@
+#include <memory>
+
+using std::string;
+using std::unordered_map;
+using std::unordered_set;
+using std::vector;
+
+extern sai_object_id_t gSwitchId;
+extern PortsOrch *gPortsOrch;
+
+static const unordered_map<string, CounterType> flex_counter_type_lookup = {
+    { PORT_INGRESS_DROPS, CounterType::PORT_DEBUG },
+    { PORT_EGRESS_DROPS, CounterType::PORT_DEBUG },
+    { SWITCH_INGRESS_DROPS, CounterType::SWITCH_DEBUG },
+    { SWITCH_EGRESS_DROPS, CounterType::SWITCH_DEBUG },
+};
+
+// Initializing DebugCounterOrch creates a group entry in FLEX_COUNTER_DB, so this
+// object should only be initialized once.
+DebugCounterOrch::DebugCounterOrch(DBConnector *db, const vector<string>& table_names, int poll_interval) :
+        Orch(db, table_names),
+        flex_counter_manager(DEBUG_COUNTER_FLEX_COUNTER_GROUP, StatsMode::READ, poll_interval),
+        m_stateDb(new DBConnector(STATE_DB, DBConnector::DEFAULT_UNIXSOCKET, 0)),
+        m_debugCapabilitiesTable(new Table(m_stateDb.get(), STATE_DEBUG_COUNTER_CAPABILITIES_NAME)),
+        m_countersDb(new DBConnector(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0)),
+        m_counterNameToPortStatMap(new Table(m_countersDb.get(), COUNTERS_DEBUG_NAME_PORT_STAT_MAP)),
+        m_counterNameToSwitchStatMap(new Table(m_countersDb.get(), COUNTERS_DEBUG_NAME_SWITCH_STAT_MAP))
+{
+    SWSS_LOG_ENTER();
+    publishDropCounterCapabilities();
+    flex_counter_manager.enableFlexCounterGroup();
+}
+
+DebugCounterOrch::~DebugCounterOrch(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+// doTask processes updates from the consumer and modifies the state of the
+// following components:
+//     1) The ASIC, by creating, modifying, and deleting debug counters
+//     2) syncd, by creating, modifying, and deleting flex counters to
+//        keep track of the debug counters
+//
+// Updates can fail due to the following:
+//    1) Malformed requests: if the update contains an unknown or unsupported
+//       counter type or drop reason then the update will fail
+//    2) SAI failures: if the SAI returns an error for any reason then the
+//       update will fail
+//  It is guaranteed that failed updates will not modify the state of the
+//  system.
+//
+// In addition, updates are idempotent - repeating the same request any number
+// of times will always result in the same external behavior.
+void DebugCounterOrch::doTask(Consumer& consumer)
+{
+    SWSS_LOG_ENTER();
+
+    // Currently we depend on 1) the switch and 2) the ports being ready
+    // before we can set up the counters. If debug counters for other
+    // object types are added we may need to update this dependency.
+    if (!gPortsOrch->allPortsReady())
+    {
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+        std::vector<FieldValueTuple> fvt = kfvFieldsValues(t);
```

Comment:
fvt seems a bit out of league compare to key/op variables naming. Consider val instead?